refactor(performance): Have Ships cache commonly used attributes and calculations

Amazinite opened the pull request:
**Refactor**

Merry Christmas! Here is my present to you.

## Acknowledgement

- [x] I acknowledge that I have read and understand the [Contributing](https://github.com/endless-sky/endless-sky/blob/master/docs/CONTRIBUTING.md) article.

## Summary

This PR refactors the Ship class to no longer access its attributes every game tick using Outfit::Get on the `attributes` variable. This resurrects my old work from #6804, resolving #6576, and also addressed one of the major performance bottlenecks outlined in #6329, that being the large amount of time spent on string comparisons.

A Ship's frequently-used attributes, as well as values that can be calculated from these attributes, are now stored in a ShipAttributesHandler class. This class is updated whenever a ship's outfits change (including when ammo is fired), storing commonly used attributes into dedicated variables and therefore only running Outfit::Get once per attribute per outfit update instead of at least once per attribute per game tick. Some attributes, such as a ship's fuel capacity or various scanning attributes, could be called numerous times per game tick, so for those attributes this is an even bigger improvement.

Almost all of the Outfit::Get calls made within Ship and by classes that interacted with Ship have been replaced to instead use the ShipAttributesHandler. The remaining exceptions include:
- UI code that only looks at one or two ships at a time.
- The attribute is only ever used by the player's flagship.
- The attribute is only called on very infrequent occasions (e.g. only called when the ship is spawned in or when it is scanned).

## Testing Done + Save File

In this save file, I have 2,500 Sparrows as escorts. I flew around various systems and engaged in combat with random merchants and Navy ships. Didn't notice any issues in the shops or with damage being dealt.
[Test Pilot.txt](https://github.com/user-attachments/files/24340119/Test.Pilot.txt)

On the master branch, I experienced a peak CPU readout of about 225%.
On #12019, my peak CPU readout increased to about 270%. (A 20% decrease in performance.)
On this PR + #12019, my peak CPU readout was only about 150%. (A 33% increase in performance over the master branch, and a 44% improvement over #12019.)

## Performance Impact

MUCH better.

Amazinite committed:
fix(mechanics): Store Outfit attributes as integers to avoid precision errors when selling large numbers of outfits

Amazinite committed:
Feedback from Tibetiroka + fix weapon ports

Amazinite committed:
Please Microsoft

Amazinite committed:
Remove std prefixes

Amazinite committed:
Update both Dictionaries when calling Outfit::Set

Amazinite committed:
Don't cache the double values for attributes within Outfit

Performance improvements will come later

Amazinite committed:
Move the management of a ship's resources and various attributes to separate classes

A ship's resources (shields, hull, energy, heat, fuel, DoT) are now stored in a ResourceLevels class. This object can also be used to store the cost of using outfits or the damage that a weapon deals to a ship/

ShipAttributeHandler caches attributes and calculations that can currently be accessed/calculated every tick in Ship, only updating them once the underlying attributes update. Currently only covers what my original PR covered.

Already resulting in a noticeable performance improvement. More to come later.

Amazinite committed:
Have ShipAttributeHandler save pointers to parts of the Ship for easier access

Amazinite committed:
Remove the wildcard field from ResourceLevels since it's too opaque

Just have extra variables on the attribute handler with easily understood names.

Amazinite committed:
Update all Outfit::Get calls in Ship (that happen after FinishLoading) to instead use the ShipAttributeHandler

Also moved jump and scram speed into ShipJumpNavigation

Amazinite committed:
Update AI to use the ShipAttributeHandler where possible

Amazinite committed:
Update Engine

Amazinite committed:
Update various other classes

Any remaining instances of a ship's attributes being accessed directly are because:
- It's UI code that only looks at one or two ships at a time
- It's only ever the player's flagship that uses the attribute
- It's a very infrequently called function (e.g. only called when the ship is spawned in or when it is scanned)

Amazinite committed:
Fix purchasing new ships

Amazinite committed:
Pretty this up

Amazinite committed:
Style on 'en

Amazinite committed:
Copy and paste moment

Amazinite committed:
Fix ion resistance

The Heron was timing out because it got hit with an ion storm and lost all its fuel due to its ion resistance

Amazinite committed:
Replace the Outfit::AttributeNames function with iterator functions over the attribute names and values

Amazinite committed:
Merge branch 'outfit-attribute-precision' into energy-handler-reborn

Amazinite committed:
Decouple from the Outfit refactor

These two don't strictly need to be together. I just started from the other PR expecting that what I was about to write would require stuff from it.

Amazinite committed:
Oops

Amazinite committed:
Oops x2

Amazinite committed:
Missed a spot

nickshanks commented on source/Ship.cpp:
```diff
@@ -193,16 +129,16 @@ namespace {
-		for(auto &&item : stockpile)
+		for(const auto &[outfit, count] : stockpile)
 		{
 			assert(item.second > 0 && "stockpile count must be positive");
```
compile error:

```suggestion
			assert(count > 0 && "stockpile count must be positive");
```

nickshanks reviewed (approved):
After merging upstream/master to be able to build on my Mac, and fixing the below, I played for about 45 minutes, mostly battling in Korath space with big JD-equipped human/pug/hai fleets, without issues. I didn't feel like I noticed considerably less sluggishness than usual, maybe a little, but it definetly wasn't the worst I've seen.
No crashes, at least.

Amazinite committed:
Curious

Co-authored-by: Nicholas Shanks <nicholas.shanks@bbc.co.uk>

Amazinite committed:
Merge remote-tracking branch 'upstream/master' into energy-handler-reborn

knarrff commented:
I ran my current test (longer automatic integration test involving a large own fleet and multiple factions battling for some time). The code of the PR is measurably faster (~20%: integration tests are run as fast as possible and are not limited to 60 fps). It also seems to run exactly the same "replay" of that game with and without the PR (but remember that this is including other changes I made to get some repeatability). So in short: from a functional point I see the benefit and don't see adverse effects. Great!

On a code-review level I cannot help but wonder if this couldn't be done in some more transparent way for users of the classes. It doesn't feel right that users/callers of class Ship should even know about the ShipAttributeHandler. I probably haven't thought it through and I certainly do not have the broad knowledge of the code base. I wonder why at least a few of these are implemented as "attritbues" at all, and not just as plain numbers within the ship class in the first place - at least for those that are generally available on ships or could be set to sentinel values indicating "that isn't an attribute this ship has". Couldn't make this the code so much simpler? As it is now, it looks like we now have a new abstraction level (the attribute handler) to fix the performance problems of the still existing abstraction level (the attributes) for "a lot of doubles" (the actual values).

Amazinite committed:
Merge remote-tracking branch 'upstream/master' into energy-handler-reborn

Amazinite committed:
Merge remote-tracking branch 'upstream/master' into energy-handler-reborn

Amazinite committed:
Move some information and functions from Ship/ShipAttributeHandler to Entity

With the introduction of Entity, I can't just get the ShipAttributeHandler from a Minable in Projectile, and I don't think it makes sense to give an attribute handler class to Minable.

Therefore, I've broken various functions out of Ship and ShipAttributeHandler and moved them into Entity to retain the same level of caching necessary for Projectile. This also means that the only time that Projectile needs to know if what it's targeting is a ship is when it's determining if it should break its target on disabled fighters.

Amazinite committed:
Oops

Amazinite commented on source/ship/ShipAttributeHandler.cpp:
```diff
@@ -0,0 +1,992 @@
+
+void ShipAttributeHandler::Capacity()
+{
+	capacity.hull = attributes->Get("hull") * (1 + attributes->Get("hull multiplier"));
```
I understand the concern, but I think that with #12019, there's really no avoiding having "two sources of truth." For any attribute stored as an integer, we need to cache the double-conversion of that attribute, lest we be constantly hit with an extra division on every attribute retrieval (which, from what I can tell from testing that PR, has a noticeable impact on performance). If we store more attributes as their own fields outside of the dictionary like we do with mass being a separate field on Outfit, then we'll just be running into #3639 again.

Amazinite committed:
Move fuel/energy capacity functions from the attribute handler to Entity

Amazinite committed:
Use FuelLevel in cases that were calculating the fuel level from the fuel fraction and capacity

Amazinite committed:
Housekeeping

Amazinite committed:
Pull some changes from the outfit generators branch

Amazinite committed:
Fix that

knarrff commented:
> @knarrff Can you contribute that integration test harness (I presume it locks the random seed and automates player actions?) as a PR. It would be good to have a consistent way to test potential performance optimisation.

Parts (the important ones) are now here: #12257

Amazinite committed:
Merge remote-tracking branch 'upstream/master' into energy-handler-reborn

Amazinite committed:
Merge fix

TheGiraffe3 commented:
conflicts

Amazinite committed:
Merge remote-tracking branch 'upstream/master' into energy-handler-reborn

Amazinite committed:
Merge fixes

Amazinite committed:
Mandatory crew attribute on ships

Amazinite committed:
Typo fix

Amazinite committed:
Move CreateSparks and related functionality to Entity

Amazinite committed:
Move status effect management to its own class that Entity uses

Amazinite committed:
Oh yeah, I renamed it

Amazinite committed:
Another one

Amazinite committed:
I renamed it twice, actually

Amazinite committed:
Housekeeping

- Removed a few imports that are now unused.
- Removed some code left over from a merge conflict.
- Added "Fraction" or "Level" to some entity functions where appropriate to disambiguate what is being returned.

Amazinite committed:
I disagree, style checker

Amazinite commented:
Marking as a draft while I break the Entity-level changes out into its own PR. Those changes can stand on their own and were made here because it simplified some of the implementation for ship attribute caching.
